### PR TITLE
Fix inf value reduction in non persistent reduction for scans

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132293

Fixes https://github.com/pytorch/pytorch/issues/132107

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -2236,19 +2236,57 @@ def cse_multiple(line, n, masks):
...
+                    return f"{var}.to({triton_compute_type(dtype)}, bitcast=True)"
+                else:
+                    return str(var)
+
             sum_helper_fn = self._lift_helper(sum_fn, len(values))
             pre_reduce_vars = ", ".join(
-                f"{scan_var} * (rbase == (RBLOCK - 1))"
-                for scan_var in partial_scan_vars
+                f"{bitcast_if_float(scan_var, dtype)} * (rbase == (RBLOCK - 1))"
+                for scan_var, dtype in zip(partial_scan_vars, dtypes)
```
Comment (peterbell10) [inline torch/_inductor/codegen/triton.py:2270]: I would suggest using a helper function:
```python
def select_one(x, mask, dim, keepdims=False):
    idtype = tl.core.get_int_dtype(x.dtype.primitive_bitwidth, signed=False)
    ix = x.to(idtype, bitcast=True)
    iy = tl.sum(x * mask, dim, keepdims=keepdims)
    return iy.to(x.dtype, bitcast=True)
```

**Comment by isuruf** [comment]
test failure seems unrelated. See https://github.com/pytorch/pytorch/pull/132468 where the same test fails.

**Review by peterbell10 — APPROVED** [review]

**Pull request closed without merging**